security/pfSense-pkg-stunnel/files/usr/local/pkg/stunnel.inc

```diff
@@ -0,0 +1,133 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+require_once("config.inc");
+require_once("util.inc");
+require_once("pfsense-utils.inc");
+require_once("certs.inc");
+
+global $config;
+define('STUNNEL_LOCALBASE', '/usr/local');
+define('STUNNEL_ETCDIR', STUNNEL_LOCALBASE . "/etc/stunnel");
+
+// Initialize the list of tunnels.
+$tunnels = $config['installedpackages']['stunnel']['config'];
+is_array($tunnels) ? $num_tunnels = count($tunnels) : $num_tunnels = 0;
+if (!isset($_GET['id']) and $num_tunnels) {
+	for ($i = 0; $i < $num_tunnels; $i++) {
+		$tunnel = $tunnels[$i];
+	}
+}
+
+function stunnel_save($config) {
+	conf_mount_rw();
+	safe_mkdir(STUNNEL_ETCDIR, 0755);
+	$fout = fopen(STUNNEL_ETCDIR . "/stunnel.conf", "w");
+	fwrite($fout, "cert = " . STUNNEL_ETCDIR . "/stunnel.pem \n");
+	fwrite($fout, "chroot = /var/tmp/stunnel \n");
+	fwrite($fout, "setuid = stunnel \n");
+	fwrite($fout, "setgid = stunnel \n");
+	if (!is_array($config['installedpackages']['stunnel']['config'])) {
+		$config['installedpackages']['stunnel']['config'] = array();
+	}
+
+	// Keep array of in-use certificates so we can clean up after ourselves.
+	$in_use_certs = array();
+	$in_use_certs[] = 'stunnel.pem';
+
+	foreach ($config['installedpackages']['stunnel']['config'] as $pkgconfig) {
+		fwrite($fout, "\n[" . $pkgconfig['description'] . "]\n");
+		if ($pkgconfig['client']) {
+			fwrite($fout, "client = yes" . "\n");
+		}
+		if ($pkgconfig['certificate'] && $pkgconfig['certificate'] != 'default') {
+			$cert = lookup_cert($pkgconfig['certificate']);
+			if ($cert != false) {
+				file_put_contents(STUNNEL_ETCDIR . "/{$pkgconfig['certificate']}.pem", 
+					base64_decode($cert['prv']) . base64_decode($cert['crt']) . ca_chain($cert));
+				fwrite($fout, "cert = " . STUNNEL_ETCDIR . "/{$pkgconfig['certificate']}.pem\n");
+				$in_use_certs[] = "{$pkgconfig['certificate']}.pem";
+			}
+		}
+		if ($pkgconfig['sourceip']) {
+			fwrite($fout, "local = " . $pkgconfig['sourceip'] . "\n");
+		}
+		fwrite($fout, "accept = " . ($pkgconfig['localip'] ? $pkgconfig['localip'] . ":" : "")  . $pkgconfig['localport'] . "\n");
+		fwrite($fout, "connect = " . $pkgconfig['redirectip'] . ":" . $pkgconfig['redirectport'] . "\n");
+		fwrite($fout, "TIMEOUTclose = 0\n\n");
+	}
+	fclose($fout);
+
+	// Clean up certs that are no longer in use.
+	foreach (glob(STUNNEL_ETCDIR . "/*.pem") as $file) {
+		if (!in_array(basename($file), $in_use_certs)) {
+			unlink($file);
+		}
+	}
+
+	conf_mount_ro();
+	stop_service("stunnel");
+	start_service("stunnel");
+}
+
+function stunnel_install() {
+	safe_mkdir(STUNNEL_ETCDIR);
+
+	// Generate a self-signed default certificate.
```

Comment:
You should check here to make sure such a cert does not already exist, otherwise any time the package is upgraded or reinstalled, it will make another similar certificate.